### DISCUSSION TITLE
'NetboxDcimModule' object has no attribute 'module'

### DISCUSSION THREAD

**Discussion by dragonfly-net** [netbox-community/ansible_modules#915] [Q&A] [answered]
### Ansible NetBox Collection version

v3.10.0

### Ansible version

ansible 2.9.27
  config file = /home/user/flora/ops/ansible_v4_htr.master/playbooks/vms/core-htr/ansible.cfg
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/flora/venv3_2.9.27/lib/python3.8/site-packages/ansible
  executable location = /home/user/flora/venv3_2.9.27/bin/ansible
  python version = 3.8.10 (default, Nov 14 2022, 12:59:47) [GCC 9.4.0]

### NetBox version

v3.4.0

### Python version

3.8

### Steps to Reproduce

ansible-galaxy collection install -f netbox.netbox

      - name: Create device within NetBox with only required information (add to cluster)
        delegate_to: 127.0.0.1
        netbox.netbox.netbox_device:
          netbox_url: http://demo.netbox.dev
          netbox_token: xxx
          data:
            name: "{{ server }}"
            cluster: "{{ server }}"
          state: present

### Expected Behavior

link cluster to device

FIx:
ansible-galaxy collection install -f netbox.netbox:==3.9.0
and all works

### Observed Behavior

TASK [Create device within NetBox with only required information (add to cluster)] ************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'NetboxDcimModule' object has no attribute 'module'
fatal: [server.loc -> 127.0.0.1]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/home/user/.ansible/tmp/ansible-tmp-1672170375.1104395-3072148-221936447477935/AnsiballZ_netbox_device.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/home/user/.ansible/tmp/ansible-tmp-1672170375.1104395-3072148-221936447477935/AnsiballZ_netbox_device.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/user/.ansible/tmp/ansible-tmp-1672170375.1104395-3072148-221936447477935/AnsiballZ_netbox_device.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.netbox.netbox.plugins.modules.netbox_device', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib/python3.8/runpy.py\", line 207, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib/python3.8/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib/python3.8/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_netbox.netbox.netbox_device_payload_exn0o1nu/ansible_netbox.netbox.netbox_device_payload.zip/ansible_collections/netbox/netbox/plugins/modules/netbox_device.py\", line 336, in <module>\n  File \"/tmp/ansible_netbox.netbox.netbox_device_payload_exn0o1nu/ansible_netbox.netbox.netbox_device_payload.zip/ansible_collections/netbox/netbox/plugins/modules/netbox_device.py\", line 331, in main\n  File \"/tmp/ansible_netbox.netbox.netbox_device_payload_exn0o1nu/ansible_netbox.netbox.netbox_device_payload.zip/ansible_collections/netbox/netbox/plugins/module_utils/netbox_dcim.py\", line 65, in __init__\nAttributeError: 'NetboxDcimModule' object has no attribute 'module'\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}

**Comment by sc68cal** [comment]
Remove the `delegate_to` and consult our documentation on how to correctly invoke these modules.

https://netbox-ansible-collection.readthedocs.io/en/latest/getting_started/how-to-use/modules.html#specifying-modules-in-a-playbook

**Reply by dragonfly-net** [reply]
1) we can't use hosts:localhost, because i need to load exist machines, with own group_vars, host_vars - and i must use delegate_to. Localhost is for some sandbox for 2 machines, not for real world
2) 3.9.0 not broken
3) this will be part of the deployment of vm's, to external devices

Now all our envs includes 6000+ip's, 600+ hypervisors, and 2000+ vms

**Reply by dragonfly-net** [reply]
I insist it's a bug

**Reply by sc68cal** [reply]
The problem is `delegate_to` has some sharp edges when it comes to python interpreter discovery and library resolution. I think your issue is similar to others where it is using the system python and libraries, and not your environment installed at `/home/user/flora/venv3_2.9.27`

**Reply by dragonfly-net** [reply]
But why all ok with 3.9.0? We freezed our versions now, but it's not very good

**Reply by dragonfly-net** [reply]
and - task breaks exact in 'device', but not in first part - even with 3.10.0 - all works
```      - name: Create cluster within NetBox with only required information
        delegate_to: 127.0.0.1
        netbox.netbox.netbox_cluster:
          netbox_url: http://demo.netbox.dev
          netbox_token: xxx
          data:
            name: "{{ server }}"
            cluster_type: "kvm-lxc"
            tags:
              - Hypervisor
          state: present
```
netbox.netbox.netbox_virtual_machine works too

**Comment by rodvand** [comment, marked as answer]
First of all - you are using an ansible version which is outdated, and we can't promise non-breaking releases for old versions. Could you check your `pip freeze` both in the virtual environment and your system environment, and make sure packaging is installed?

**Reply by dragonfly-net** [reply]
See above: broken only work with devices, not VM and clusters, and downgrade to 3.9.0 fixes the bug
pip freeze:
```ansible==2.9.27
certifi==2022.12.7
cffi==1.15.1
charset-normalizer==2.1.1
cryptography==38.0.1
idna==3.4
Jinja2==3.1.2
MarkupSafe==2.1.1
pycparser==2.21
pynetbox==7.0.0
PyYAML==6.0
requests==2.28.1
urllib3==1.26.13
```

**Reply by rodvand** [reply]
Packaging is a required package. Please install this and check again.

**Reply by dragonfly-net** [reply]
Oh, now it's fixed, thanks! I think, need to update requirements..
